flocker/node/script.py

```diff
@@ -281,143 +319,282 @@ def get_service(self, reactor, options):
+        backend_name = api_args.pop('backend')
+
+        return cls(
+            control_service_host=host,
+            control_service_port=port,
+
+            node_credential=node_credential,
+            ca_certificate=ca_certificate,
+
+            backend_name=backend_name.decode("ascii"),
+            api_args=api_args,
         )
 
-        service_factory = AgentServiceFactory(
-            deployer_factory=deployer_factory
-        ).get_service
+    def get_backend(self):
+        """
+        Find the backend in ``self.backends`` that matches the one named by
+        ``self.backend_name``.
+        """
+        for backend in self.backends:
+            if backend.name == self.backend_name:
+                return backend
+        raise ValueError(
+            "Backend named {} not available".format(self.backend_name),
+        )
 
-        service = service_factory(reactor, options)
+    def get_tls_context(self):
+        """
+        Get some TLS configuration objects which will authenticate this node to
+        the control service node and the reverse.
+        """
+        policy = ControlServicePolicy(
+            ca_certificate=self.ca_certificate,
+            client_credential=self.node_credential.credential,
+        )
+        return _TLSContext(
+            context_factory=policy.creatorForNetloc(
+                self.control_service_host, self.control_service_port,
+            ),
+            node_credential=self.node_credential,
+        )
 
-        if configuration['dataset']['backend'] == 'zfs':
-            # XXX This should not be a special case,
-            # see https://clusterhq.atlassian.net/browse/FLOC-1924.
-            volume_service.setServiceParent(service)
+    def get_api(self):
+        """
+        Get an storage driver which can be used to create an ``IDeployer``.
 
-        return main_for_service(
-            reactor=reactor,
-            service=service,
+        :return: An object created by one of the factories in ``self.backends``
+            using the configuration from ``self.api_args`` and other useful
+            state on ``self``.
+        """
+        backend = self.get_backend()
+
+        api_args = self.api_args
+        if backend.needs_id:
+            cluster_id = self.node_credential.cluster_uuid
+            api_args = api_args.set("cluster_id", cluster_id)
+        if backend.needs_reactor:
+            api_args = api_args.set("reactor", self.reactor)
+
+        return backend.api_factory(**api_args)
+
+    def get_deployer(self, api):
+        """
+        Create an ``IDeployer`` provider suitable for the configured backend
+        and this node.
+
+        :return: The ``IDeployer`` provider.
+        """
+        backend = self.get_backend()
+        deployer_factory = self.deployers[backend.deployer_type]
+
+        hostname = self.get_external_ip(
+            self.control_service_host, self.control_service_port,
+        )
```

Comment:
:question: Perhaps we should stop saying hostname, where possible. This variable will certainly be an IPv4 address.